sycl/doc/Assert.md

```diff
@@ -0,0 +1,421 @@
+fallback implementation is used.
+
+
+## Safe approach
+
+This is the preferred approach and implementations should use it when possible.
+It guarantees assertion failure notification delivery to the host regardless of
+kernel behavior which hit the assertion.
+
+The Native Device Compiler is responsible for providing implementation of
+`__devicelib_assert_fail` which completely hides details of communication
+between the device code and the Low-Level Runtime from the SYCL device compiler
+and runtime. The Low-Level Runtime is responsible for:
+ - detecting if assert failure took place;
+ - flushing assert message to `stderr` on host.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+      // It's not defined if GPU thread stops execution
+      // Other GPU threads are left untouched
+   2. Specialized version of `__devicelib_assert_fail` is called
+   3. Device immediately signals to host (Low-Level Runtime)
+ - Host side:
+   1. The assert failure gets detected by Low-Level Runtime
+   2. Low-Level Runtime prints assert failure message to `stderr`
+   3. Low-Level Runtime calls `abort()`
+
+
+## Fallback approach
+
+If Device-side Runtime doesn't support `__devicelib_assert_fail` then a fallback
+approach comes in place. The approach doesn't require any support from
+Device-side Runtime and Native Device Compiler. Neither it does from Low-level
+Runtime.
+
+Within this approach, a mutable program scope variable is introduced. This
+variable stores a flag which says if an assert failure was encountered. Fallback
+implementation of `__devicelib_assert_fail` atomically raises the flag so that
+DPCPP Runtime is able to detect assert failure after kernel finishes.
+
+The following sequence of events describes how user code gets notified:
+ - Device side:
+   1. Assert fails in device-code in kernel
+   2. Fallback version of `__devicelib_assert_fail` is called
+   3. Assert information is stored into program-scope variable
+   4. Kernel continues running
+ - Host side:
+   1. A copy 'kernel' is enqueued as the one depending on user's kernel to get
+      the value of assert failure flag.
+   2. A host-task is enqueued to check value of assert failure flag.
+   3. The host task calls abort whenever assert failure flag is set.
+
+Illustrating this with an example, lets assume the user enqueues three kernels:
+ - `Kernel #1`
+ - `Kernel #2`
+ - `Kernel #3`, which depends on `Kernel #1`
+
+The resulting graph will look like this: ![graph](images/assert-fallback-graph.svg)
+
+### Interface to program scope variable
+
+Multiple translation units could be compiled/linked into a single device binary
+image. All of them should have `extern` declaration of program scope variable
+available. Definition of the variable is only available within devicelib in the
+same binary image where fallback `__devicelib_assert_fail` resides.
```

Comment:
Is this unconditional or only when there is no NDEBUG? Please clarify in the text.

Reply:
Fixed.